lib/registration.cpp

```diff
@@ -13,4 +20,309 @@ transform_pointcloud_inplace(EigenPointCloudRef cloud,
+        subsets_[root_j] = root_i;
+
+        return root_i;
+      }
+    } else {
+      // Always merge i's subset into j's subset
+      numbers_[root_j] += numbers_[root_i];
+      subsets_[root_i] = root_j;
+      numbers_[root_i] = 0;
+      return root_j;
+    }
+  }
+
+  // If the subsets already have the same root representative, no action is
+  // needed
+  return root_i;
+}
+
+std::size_t
+estimate_supervoxel_count(EigenPointCloudConstRef cloud, double seed_resolution)
+{
+  // Calculate the point cloud point bounding box
+  using Coordinate = Eigen::Array<double, 1, 3>;
+  Coordinate lowerleft = cloud.colwise().minCoeff();
+  Coordinate upperright = cloud.colwise().maxCoeff();
+
+  // Subdivide the bounding box according to the given seed_resolution
+  using std::floor;
+  using MultiIndex = Eigen::Array<std::size_t, 1, 3>;
+  MultiIndex voxelsizes =
+    floor(((upperright - lowerleft) / seed_resolution) + 1)
+      .cast<std::size_t>()
+      .eval();
+
+  // Assign all points to their voxels by throwing them in a hashmap
+  std::unordered_map<std::size_t, std::vector<Eigen::Index>> voxelmap;
+  for (IndexType i = 0; i < cloud.rows(); ++i) {
+    Coordinate coord(cloud.row(i));
+    auto ind = floor(((coord - lowerleft) / seed_resolution) + 1)
+                 .cast<std::size_t>()
+                 .eval();
+    voxelmap[ind[2] * voxelsizes[1] * voxelsizes[0] + ind[1] * voxelsizes[0] +
+             ind[0]]
+      .push_back(i);
+  }
+
+  return voxelmap.size();
+}
+
+double
+squaredEuclideanDistance(const Eigen::RowVector3d& point1,
+                         const Eigen::RowVector3d& point2)
+{
+  const auto diff = point1 - point2;
+  return diff.squaredNorm();
+}
+
+std::vector<std::vector<int>>
+supervoxel_segmentation(Epoch& epoch,
+                        const KDTree& kdtree, //?
+                        double resolution,
+                        int k)
+{
+
+  // Define number of supervoxels and labels.
+  auto n_supervoxels = estimate_supervoxel_count(epoch.cloud, resolution);
+  std::vector<int> labels(epoch.cloud.rows(), -1);
+  DisjointSet set(epoch.cloud.rows());
+
+  // Calculate normals for vccs metric
+  epoch.kdtree.build_tree(10);
+  std::vector<double> normal_radii{ 2.0 };
+  EigenNormalSet normals(epoch.cloud.rows(), 3);
+  EigenNormalSet orientation(1, 3);
+  orientation << 0, 0, 1;
+  compute_multiscale_directions(
+    epoch, epoch.cloud, normal_radii, orientation, normals);
+
+  normals.norm();
+
+  // Calculate k neigbors and its distances for each point
+  KDTree::NearestNeighborsDistanceResult result;
+  kdtree.nearest_neighbors_with_distances(epoch.cloud, result, k);
+
+  int supervoxels_amount = epoch.cloud.rows();
+
+  // calculate lambda for segmentation
+  DistanceVector lambda_distances;
+  for (size_t i = 0; i < result.size(); ++i) {
+    std::nth_element(result[i].second.begin(),
+                     result[i].second.begin() + 1,
+                     result[i].second.end());
```

Comment:
I am confused by this. Arent the KNN results already sorted by distance? IIRC, the radius search results are or that was a configurable parameter of `nanoflann`.